File: projects/dejavu.rs/dejavu-ir/src/normalize.rs
use serde_json::{Map, Value};

/// Normalize an IR JSON value for semantic equality.
pub fn normalize_value(value: Value) -> Value {
    match value {
        Value::Object(map) => {
            let mut out = Map::new();
            for (k, v) in map {
                if k == "span" {
                    continue;
                }
                let nv = normalize_value(v);
                if k == "raw" && nv == Value::Bool(false) {
                    continue;
                }
                if k == "type" {
                    // keep type
                }
                if let (Some(Value::String(t)), Value::String(text)) =
                    (out.get("type").cloned().or_else(|| if k == "type" { Some(nv.clone()) } else { None }), &nv)
                {
                    if t == "Text" && k == "value" && text.is_empty() {
                        return Value::Null;
                    }
                }
                out.insert(k, nv);
            }
            // Drop empty Text objects
            if out.get("type") == Some(&Value::String("Text".into())) && out.get("value") == Some(&Value::String(String::new()))
            {
                return Value::Null;
            }
            Value::Object(sort_keys(out))
        }
        Value::Array(items) => Value::Array(items.into_iter().map(normalize_value).filter(|v| !v.is_null()).collect()),
        other => other,
    }
}

fn sort_keys(map: Map<String, Value>) -> Map<String, Value> {
    let mut keys: Vec<_> = map.keys().cloned().collect();
    keys.sort();
    let mut sorted = Map::new();
    for k in keys {
        if let Some(v) = map.get(&k) {
            sorted.insert(k, v.clone());
        }
    }
    sorted
}

File: projects/dejavu.ts/dejavu-types/src/normalize.ts
import type {IrValue} from "./ir";

/** Normalize IR JSON for semantic equality (see specifications/ir/v1/normalize.md). */
export function normalizeValue(value: unknown): unknown {
    if (Array.isArray(value)) {
        return value.map(normalizeValue).filter((v) => v !== null);
    }
    if (value && typeof value === "object") {
        const obj = value as Record<string, unknown>;
        const out: Record<string, unknown> = {};
        for (const key of Object.keys(obj).sort()) {
            if (key === "span") continue;
            if (key === "raw" && obj[key] === false) continue;
            const nv = normalizeValue(obj[key]);
            out[key] = nv;
        }
        if (out.type === "Text" && out.value === "") {
            return null;
        }
        return out;
    }
    return value as IrValue;
}

export function normalizeIrJson(json: string): unknown {
    return normalizeValue(JSON.parse(json));
}
